### ISSUE TITLE
Electron version hangs quite randomly at startup, with "Network service crashed" error message

### ISSUE THREAD

**Issue by francisbarton** [rstudio/rstudio#12469]
### System details

    RStudio Edition : Desktop
    RStudio Version : RStudio 2022.12.0+353 "Elsbeth Geranium" Release (7d165dcfc1b6d300eb247738db2c7076234f6ef0, 2022-12-03) for Windows Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) RStudio/2022.12.0+353 Chrome/102.0.5005.167 Electron/19.1.3 Safari/537.36
    OS Version      : Windows 10 Enterprise Version 20H2 Build 19042.2251
    R Version       : R version 4.2.2 (2022-10-31 ucrt) -- "Innocent and Trusting"

### Steps to reproduce the problem

1. Open RStudio.
2. Open a project.
3. Open another project using projects dropdown, "Open project in a new R session"
4. Repeat 3 (you now have three RStudio projects open in three windows)
5. Repeat 3 - this session/window does not render.

Alternative step 5:
* Click Session (menu) > New Session (has same result)

### Describe the problem in detail

On trying to open a fourth window/session/project, RStudio hangs at the splash screen.
The three open projects remain functional - it's not a general crash.

I have just installed Elsbeth Geranium stable after being on Spotted Wakerobin stable.
I never had a problem with opening >3 sessions previously.
I don't think I was on an Electron build though? - May be related to that.
Or a memory problem on my machine - it is quite limited.
Possibly related to #11604 .

I have tried closing one of the three open projects, and then opening the "fourth" project as the third project (does this make sense!?) - this works.
So now have three projects open OK. (i.e. it's not to do with the fourth project itself.)
Then if I try to open a fourth project, I get the hang at splash screen again.

### Describe the behaviour you expected

Fourth project/window/session would launch successfully.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [ ] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by kevinushey** [comment]
To confirm -- does the 4th instance hang forever, or do you see it open "eventually", e.g. after a minute or so? (We had seen some reports of this in the past but want to confirm if it might be worse than we initially concluded.)

**Comment by francisbarton** [comment]
It seems to hang at the "R" logo splash screen stage indefinitely. Several minutes (at least 10), and more. I haven't yet left it to see if eventually opens. But I can run an experiment now just to confirm.

**Comment by francisbarton** [comment]
I just explored this further by reverting to Spotted Wakerobin (2022.07.2-576) Electron and "normal" (Qt?) and trying to replicate.

With the Electron version, I could open 3 simultaneous sessions but a fourth session launch did not succeed. (No splash screen in this version.)

With the Qt version, I could open >3 instances no problem. So perhaps it is Electron-related.

**Comment by ronblum** [comment]
I tried reproducing this on Windows 11 both by open six projects in new sessions, as well as doing the same with just creating new sessions. No luck.

@kevinushey Is there any system information that would be useful to share, to compare @francisbarton and my hardware and OS?

**Comment by francisbarton** [comment]
I also tried going "forward" to RStudio-2022.12.0-354 but that wouldn't load for me at all - just a hang at the splash screen upon opening a single instance! Very weird.

**Comment by kevinushey** [comment]
@francisbarton, does it by any chance depend on whether you're opening the same project multiple times in different windows, versus different projects in each window?

Or, does it depend on which window you're attempting to open the new project from (the "main" window versus one of the newly-launched windows)?

Are you opening the projects all at once, or opening then sequentially after each is opened?

Ultimately, I wasn't able to reproduce either :-/ so there must be some other variable we're missing.

**Comment by erex** [comment]
I don't know if this is related to OP, RStudio 2022.12.0 Build 353 hangs when I attempt to close a project (just a single project). The project will not close, nor will RStudio close; I must "End Task" using Task Manager. Problem did not exist prior to my upgrade of RStudio.

**Comment by francisbarton** [comment]
@kevinushey apologies for delay in replying, it's my work machine that has the problem and it's been enjoying a holiday for the last week or so 🎅🏻😄
I'm not able to reproduce the issue at the moment in order to answer your question fully. At the moment I am finding that I can't even open a _second_ instance without it stalling, which is odd because before it very clearly instance 4 that was causing the issue.
When I posted the issue originally, I think I was opening window b from a, window c from b, window d from c. Each with a different project/session. But I'm not 100% sure. Wish I could test it properly again.

**Comment by francisbarton** [comment]
Further info on trying to run 2022.12.0-354 Electron. On trying to run the app for the first time (ie just a single initial instance) I get an indefinite hang at the RStudio splash screen.
I am not sure if the process screenshot attached tells you anything. (Ignore the 100% Disk - I think that is just a temporary system scan going on in the background).
![image](https://user-images.githubusercontent.com/1819920/210395447-9187bef1-0665-46bb-b096-51acbd12dae8.png)

**Comment by francisbarton** [comment]
Latest investigation:
Back to version 2022.07.2+576 Chrome/102.0.5005.63 Electron/19.0.4 Safari/537.36

First instance, window A, opens. From A, I can launch instance B and then instance C (each separate R projects) successfully.
Then further attempts to launch new projects/instances do not succeed. No new windows are created.

If you look at the screenshot below, you can see instances B and C each with 8 processes. Instance A accumulates more processes each time I try to launch a new instance (I think I have tried 7 unsuccessful launches!)
![image](https://user-images.githubusercontent.com/1819920/210457857-ed61746c-7fb5-4031-b641-1624a170769e.png)

In `$HOME/AppData/Local/RStudio/log/rdesktop.2023-01-03.log` I now have
```
{"level":"error","message":"Failed to load http://127.0.0.1:40373: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:40373'"}
{"level":"error","message":"Failed to load http://127.0.0.1:43475: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:43475'"}
{"level":"error","message":"Failed to load http://127.0.0.1:17645: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:17645'"}
{"level":"error","message":"Failed to load http://127.0.0.1:46533: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:46533'"}
{"level":"error","message":"Failed to load http://127.0.0.1:22271: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:22271'"}
{"level":"error","message":"Failed to load http://127.0.0.1:24489: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:24489'"}
{"level":"error","message":"Failed to load http://127.0.0.1:25859: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:25859'"}
{"level":"error","message":"Failed to load http://127.0.0.1:8235: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:8235'"}
{"level":"error","message":"Failed to load http://127.0.0.1:23917: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:23917'"}
```

**Comment by francisbarton** [comment]
From the situation in the comment above, if I then close window C, leaving just A and B still open, and I then try to re-open C from B I get the following situation:
![image](https://user-images.githubusercontent.com/1819920/210459280-81e2e9d7-d326-453a-83e3-82a05207e96e.png)

B now has 15 processes. C fails to open. Only A and B are open.

**Comment by francisbarton** [comment]
Weird error message when closing down rogue R/RStudio processes in Task Manager:

```
2023-01-03T18:44:08.628486Z [rsession-francis.barton] ERROR system error 10053 (An established connection was aborted by the software in your host machine) [request-uri: /events/get_events]; OCCURRED AT void __cdecl rstudio::session::HttpConnectionImpl<class rstudio_boost::asio::ip::tcp>::sendResponse(const class rstudio::core::http::Response &) C:\Users\jenkins\workspace\ide\open-source-pipeline\spotted-wakerobin\src\cpp\session\http\SessionHttpConnectionImpl.hpp:156; LOGGED FROM: void __cdecl rstudio::session::HttpConnectionImpl<class rstudio_boost::asio::ip::tcp>::sendResponse(const class rstudio::core::http::Response &) C:\Users\jenkins\workspace\ide\open-source-pipeline\spotted-wakerobin\src\cpp\session\http\SessionHttpConnectionImpl.hpp:161

2023-01-03T23:51:56.294326Z [rsession-francis.barton] ERROR system error 10053 (An established connection was aborted by the software in your host machine) [request-uri: /events/get_events]; OCCURRED AT void __cdecl rstudio::session::HttpConnectionImpl<class rstudio_boost::asio::ip::tcp>::sendResponse(const class rstudio::core::http::Response &) C:\Users\jenkins\workspace\ide\open-source-pipeline\spotted-wakerobin\src\cpp\session\http\SessionHttpConnectionImpl.hpp:156; LOGGED FROM: void __cdecl rstudio::session::HttpConnectionImpl<class rstudio_boost::asio::ip::tcp>::sendResponse(const class rstudio::core::http::Response &) C:\Users\jenkins\workspace\ide\open-source-pipeline\spotted-wakerobin\src\cpp\session\http\SessionHttpConnectionImpl.hpp:161

2023-01-03T23:51:56.294326Z [rsession-francis.barton] ERROR system error 10053 (An established connection was aborted by the software in your host machine) [request-uri: /events/get_events]; OCCURRED AT void __cdecl rstudio::session::HttpConnectionImpl<class rstudio_boost::asio::ip::tcp>::sendResponse(const class rstudio::core::http::Response &) C:\Users\jenkins\workspace\ide\open-source-pipeline\spotted-wakerobin\src\cpp\session\http\SessionHttpConnectionImpl.hpp:156; LOGGED FROM: void __cdecl rstudio::session::HttpConnectionImpl<class rstudio_boost::asio::ip::tcp>::sendResponse(const class rstudio::core::http::Response &) C:\Users\jenkins\workspace\ide\open-source-pipeline\spotted-wakerobin\src\cpp\session\http\SessionHttpConnectionImpl.hpp:161
```

There is no C:\Users\jenkins on my machine.

**Comment by kevinushey** [comment]
> 2023-01-03T16:54:23.433133Z [rsession-francis.barton] ERROR Unable to retrieve user home path. HRESULT:  0x80070003; LOGGED FROM: class rstudio::core::FilePath __cdecl rstudio::core::system::userSettingsPath(const class rstudio::core::FilePath &,const class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > &,bool) src/cpp/core/system/Win32System.cpp:341 (including full error message here - have truncated above).

That error code would seem to indicate that RStudio was unable to find your home directory for some reason. (0x8007003 is the error code for "path not found").

> Interesting that it is trying different ports - is this expected?

That is expected; RStudio normally tries to use a different port number of each instance.

I suspect we're running into some kind of file / instance locking issue unique to Electron, but it's hard to know for sure... @timtmok, do you have any ideas?

**Comment by francisbarton** [comment]
Thanks @kevinushey.

To summarise:
* In RStudio 2022.07.2+576 (Qt) I can open >3 instances no problem (and quite quickly)
* In RStudio 2022.07.2+576 (Electron) I can open up to 3 instances (comparatively slowly) but not a fourth
* In RStudio 2022.12.0+353 or 2022.12.0+354 I can't open a single instance (hangs at splash).

Something to do with accessing my home directory is a likely culprit.
This is potentially related to a quirk of my work machine, where my normal home dir is under C:\Users as usual, but in some contexts a network drive (H:\) is treated as my home dir. For example, Git Bash opens in H:\ and uses H:\.ssh as its ~/.ssh folder.
I'm going to try [setting $HOME](https://stackoverflow.com/a/8562715/5168907) to my user profile instead of H:\ and see if that changes things.

**Comment by ronblum** [comment]
@francisbarton Is the third scenario above what's described in #12497?

**Comment by francisbarton** [comment]
> @francisbarton Is the third scenario above what's described in #12497?

yes, it sounds exactly the same.

**Comment by ronblum** [comment]
@francisbarton Thanks!

**Comment by timtmok** [comment]
> I suspect we're running into some kind of file / instance locking issue unique to Electron, but it's hard to know for sure... @timtmok, do you have any ideas?

Sounds like it's possible that the home directory isn't accessible when starting up. I'm not sure why that would happen in some instances. Maybe it's the project path that has an issue.

**Comment by francisbarton** [comment]
Another bit of investigation:

Process part 1:
1. Open RStudio window (session A)
2. Use Session > New Session to create new session (B)
3. Use Session > New Session to create new session (C)

We now have 3 sessions running just fine, and in Windows Task Manager each window is showing as `RStudio (4)` with its 4 processes (Console Window Host, 2 x RStudio, RStudio R Session).

Now, process part 2:
1. From session B, use Session > New Session to create new session. This does not succeed.
2. From session C, use Session > New Session to create new session. This also does not succeed.

In Windows Task Manager, each of B and C are now showing as `RStudio (11)` with 2 x Console Window Host, crashpad_handler.exe, 6 x RStudio, and 2 x RStudio R Session. A is still as `RStudio (4)`. The new session attempts spawn 7 new processes, but these "get stuck" within the existing session and do not create a new session.

Now, process part 3a:
1. Close session C. _We now only have 2 open sessions, and so there is 'space' for a third session to be launched._
2. From session B, use Session > New Session to create new session. This does not succeed.

In Windows Task Manager, session B is now showing as `RStudio (18)` with 3 x Console Window Host, 2 x crashpad_handler.exe, 10 x RStudio, and 3 x RStudio R Session. A is still as `RStudio (4)`.

Now, process part 3b:
1. From session A, use Session > New Session to create new session (D). This _succeeds_.

In Windows Task Manager, session A swells to `RStudio (12)`, but eventually gives birth to session D, which becomes visible as `RStudio (8)`, with 1 x Console Window Host, 1 x crashpad_handler.exe, 5 x RStudio, and 1 x RStudio R Session. Session A, post-birth, returns to having its original 4 processes listed (though is now still titled `RStudio (12)`).

After part 2, when only 2 sessions were open, B was not able to launch session D, because it was already clogged with the queue from the failed launch in part 2 step 1. However, at this same point, session A _was_ able to successfully launch session D, because it had not had a failed attempt in part 2.

Process part 4:
1. Close session B. Sessions A and D remain open.
2. From session A, use Session > New Session to create new session (E). This succeeds.
3. We now have 3 sessions open.
4. From session E, use Session > New Session to create new session (F). This succeeds.
5. We now have 4 sessions open!
6. From session A, use Session > New Session to create new session. This does not succeed.
7. From session D, use Session > New Session to create new session. This does not succeed.
8. From session F, use Session > New Session to create new session. This does not succeed.

**Comment by francisbarton** [comment]
Trying something out with 2022.12.0+353:

**Process 1**:

1. Open session A.
2. From session A, use Session > New Session to open session B. Succeeds.
3. From session B, use Session > New Session to open session C. Succeeds.
4. From session C, use Session > New Session to open session D. Succeeds.
5. From session D, use Session > New Session to open session E. Succeeds.
6. ... etc

OK this is progress, this version was not getting past the splash logo for me earlier today.


**Process 2**:

1. Open session A. Open an RStudio project.
2. From session A, use 'Open project in new session...' to open a project in session B. Succeeds.
3. From session B, use 'Open project in new session...' to open a project in session C. Succeeds.
4. ... etc

OK this is weird (but great!) - in Elsbeth Geranium 353, my initial issue is now not reproducible. It seems to be working fine now. I can open 4 or more sessions in Electron. I don't know what was the blocker before. Must have been Mercury in Sagittarius or something.

I can also close all those sessions as normal, without leaving any abandoned RStudio processes needing to be manually ended. No lines added to the logs either.

Shall we close this?

**Comment by francisbarton** [comment]
OK I think I am on to something. It may be a clash with VSCode. The following procedure seems to be reproducible for me (I've done this twice - will do a system reboot and then test again):

1. Open RStudio (v354 in this case). I'm using `./rstudio.exe` from a Git Bash prompt. Succeeds. No lines added to log. No VSCode running.
2. Close RStudio
3. Open VSCode (no workspace open).
4. Open RStudio. Succeeds. No lines added to log.
5. Close RStudio
6. Open a workspace in VSCode. (No need to explicitly open an R Terminal or attach R).
7. Open RStudio. Hangs at splash. Message `[23264:0105/121233.453:ERROR:network_service_instance_impl.cc(470)] Network service crashed, restarting service.` appears in Git Bash terminal. Message `2023-01-05T12:12:33.462Z ERROR Failed to load http://127.0.0.1:28955: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:28955'` is added to AppData\Local\RStudio\log\rdesktop.2023-01-05.log
8. Quit RStudio
9. Close VSCode workspace (VSCode remains open, just with no workspace active)
10. Open RStudio. Hangs at splash. Message `[21636:0105/122740.873:ERROR:network_service_instance_impl.cc(470)] Network service crashed, restarting service.` shown in Git Bash terminal. Line `2023-01-05T12:27:40.878Z ERROR Failed to load http://127.0.0.1:40701: Error: ERR_FAILED (-2) loading 'http://127.0.0.1:40701'` is added to rdesktop log.
11. Quit RStudio
12. Close VSCode completely
13. Open RStudio. Succeeds.

**Comment by ronblum** [comment]
We're still trying to figure this one out. There are now test builds with improved logging, available at https://dailies.rstudio.com/rstudio/elsbeth-geranium/electron/windows/ . The most recent is 2022.12.1-360. Please note that while the changes are minimal, this is technically a test build, but if you're willing to try it out, here's what would help:

Before running RStudio, you first need to set the logging level for debugging. The easiest way is to set `RS_LOG_LEVEL` with the value `DEBUG` as an environment variable. The log files will be at the`%localappdata%\RStudio\log` directory.
Providing any logs will be helpful, but the one in the format `rdesktop.log` is the most important.

It would still be useful to get the following information, as well:

### System Environment
    RStudio Edition     :  
    RStudio Version     :  
    OS Version          :  
    R Version           :  
    R Install Directory :  
    R Install Scope     :  
    R Registry Entry    :

**Comment by ronblum** [comment]
We've just released [RStudio version 2023.09.0+463](https://posit.co/download/rstudio-desktop/). Is this problem still occurring? If so, the logging and other information would be useful.

I'm closing this due to inactivity, but if you have any updates, please let us know!
